crates/energy/src/graph.rs

```diff
@@ -0,0 +1,259 @@
+) {
+    nearby_units
+        .par_iter_mut()
+        .for_each_mut(|(entity, mut nearby_units, transform)| {
+            if let Some(last_pos) = nearby_units.last_pos {
+                if transform.translation.to_flat().distance_squared(last_pos) < 0.5 {
+                    return;
+                }
+            }
+            nearby_units.last_pos = Some(transform.translation.to_flat());
+
+            let aabb = &Aabb::new(
+                Point::from(transform.translation - Vec3::splat(MAX_DISTANCE)),
+                Point::from(transform.translation + Vec3::splat(MAX_DISTANCE)),
+            );
+
+            let producers = spacial_index_producer.query_aabb(aabb, Some(entity));
+
+            let receivers = spacial_index_receiver.query_aabb(aabb, Some(entity));
+
+            nearby_units.units.clear();
+
+            nearby_units
+                .units
+                .producers
+                .extend(producers.map(|entity| entity.into()));
+            nearby_units
+                .units
+                .receivers
+                .extend(receivers.map(|entity| entity.into()));
+        });
+}
+
+fn update_graph(
+    mut power_grid: ResMut<PowerGrid>,
+    nearby_units: Query<(Entity, &NearbyUnits), Changed<NearbyUnits>>,
+) {
+    for (entity, nearby_units) in nearby_units.iter() {
+        let mut edges_to_remove = HashSet::new();
+
+        edges_to_remove.extend(
+            &mut power_grid
+                .graph
+                .edges(entity)
+                .map(|edge| (edge.source(), edge.target())),
+        );
+
+        let mut edges_to_add = vec![];
+
+        for nearby_entity in nearby_units.units.both() {
+            edges_to_add.push((entity, nearby_entity.into()));
+        }
+
+        for edge in edges_to_add.iter() {
+            if !power_grid.graph.contains_edge(edge.0, edge.1) {
+                power_grid.graph.add_edge(edge.0, edge.1, 1.0);
+            }
+            if edges_to_remove.contains(edge) {
+                edges_to_remove.remove(edge);
+            }
+        }
+
+        for edge in edges_to_remove.iter() {
+            power_grid.graph.remove_edge(edge.0, edge.1);
+        }
+    }
+}
+
+fn remove_old_nodes(
+    mut power_grid: ResMut<PowerGrid>,
+    mut nearby_units: Query<&mut NearbyUnits>,
+    mut death_events: EventReader<DespawnedComponentsEvent<NearbyUnits>>,
+) {
+    for event in death_events.iter() {
+        power_grid.graph.remove_node(event.entity);
+
+        // Remove the entity from the nearby units of all nearby units
+        for mut nearby_units in nearby_units.iter_mut() {
+            nearby_units.units.remove_matching(event.entity.into())
+        }
+    }
+}
+
+#[cfg(feature = "energy_graph_debug_lines")]
+fn debug_lines(
+    power_grid: Res<PowerGrid>,
+    query: Query<&Transform>,
+    mut debug_lines: ResMut<DebugLines>,
+) {
+    for (node, _) in power_grid.graph.node_references() {
```

Comment:
Doesn't this add every edge twice (since both end nodes are iterated over)? I would instead use [all_edges](https://docs.rs/petgraph/latest/petgraph/graphmap/struct.GraphMap.html#method.all_edges)